repose-aggregator/components/filters/saml-policy-translation-filter/src/main/scala/org/openrepose/filters/samlpolicy/SamlPolicyTranslationFilter.scala

```diff
@@ -189,7 +190,46 @@ class SamlPolicyTranslationFilter @Inject()(configurationService: ConfigurationS
     * @return the issuer for the embedded assertions
     * @throws SamlPolicyException if response is invalid
     */
-  def validateResponseAndGetIssuer(document: Document): String = ???
+  def validateResponseAndGetIssuer(document: Document): String = {
+    val version = 30
+    val assertionPath = "/s2p:Response/s2:Assertion"
+    val issuerPath = "/s2p:Response//s2:Issuer"
+    val signaturePath = "/s2p:Response/s2:Assertion/sig:Signature"
+    var assertionExpression : Option[XPathExpression] = None
+    var issuerExpression : Option[XPathExpression] = None
+    var signatureExpression : Option[XPathExpression] = None
+
+    try {
+      assertionExpression = Option(XPathExpressionPool.borrowExpression(assertionPath, namespaceContext, version))
+      issuerExpression = Option(XPathExpressionPool.borrowExpression(issuerPath, namespaceContext, version))
+      signatureExpression = Option(XPathExpressionPool.borrowExpression(signaturePath, namespaceContext, version))
+
+      val assertions = assertionExpression.get.evaluate(document, XPathConstants.NODESET).asInstanceOf[NodeList]
+      val issuers = issuerExpression.get.evaluate(document, XPathConstants.NODESET).asInstanceOf[NodeList]
+      val signatures = signatureExpression.get.evaluate(document, XPathConstants.NODESET).asInstanceOf[NodeList]
+
+      if (assertions.getLength != signatures.getLength) {
+        throw SamlPolicyException(400, "All assertions must be signed")
+      }
+
+      if (issuers.getLength != (assertions.getLength + 1)) {
+        throw SamlPolicyException(400, "SAML Response and all assertions need an issuer")
+      }
+
+      val listOfIssuers = for (i <- 0 until issuers.getLength) yield issuers.item(i).getTextContent
```

Comment:
Ugh... why is `issuers` not a `Collection` of `Node`'s... :disappointed:
This could be so much cleaner if they were.